systems/framework/lcm_driven_loop.h

```diff
@@ -0,0 +1,217 @@
+    name_to_input_sub_map_.insert(std::make_pair(
+        input_channel,
+        drake::lcm::Subscriber<InputMessageType>(drake_lcm_, input_channel)));
+
+    active_channel_ = input_channel;
+  };
+
+  /// Constructor for multi-input LcmDrivenLoop
+  ///     @param drake_lcm DrakeLcm
+  ///     @param diagram A Drake diagram
+  ///     @param lcm_parser The LeafSystem of the diagram that parses the
+  ///     incoming lcm message
+  ///     @param switch_channel The name of the switch channel
+  ///     @param input_channels The names of the input channels
+  ///     @param active_channel The name of the initial active input channel
+  LcmDrivenLoop(drake::lcm::DrakeLcm* drake_lcm,
+                std::unique_ptr<drake::systems::Diagram<double>> diagram,
+                const drake::systems::LeafSystem<double>* lcm_parser,
+                const std::string& switch_channel,
+                std::vector<std::string> input_channels,
+                const std::string& active_channel)
+      : LcmDrivenLoop(drake_lcm, std::move(diagram), lcm_parser) {
+    DRAKE_DEMAND(!input_channels.empty());
+    is_multiple_inputs_ = true;
+
+    // Create subscriber for the switch
+    switch_sub_ = std::make_unique<drake::lcm::Subscriber<SwitchMessageType>>(
+        drake_lcm_, switch_channel);
+
+    // Create subscribers for inputs
+    for (const auto& name : input_channels) {
+      std::cout << "Constructing subscriber for " << name << std::endl;
+      name_to_input_sub_map_.insert(std::make_pair(
+          name, drake::lcm::Subscriber<InputMessageType>(drake_lcm_, name)));
+    }
+
+    // Default initial active channel
+    active_channel_ = active_channel;
+  };
+
+  // Start simulating the diagram
+  void Simulate(double end_time = std::numeric_limits<double>::infinity()) {
+    // Get mutable contexts
+    auto& diagram_context = simulator_->get_mutable_context();
+    auto& lcm_parser_context = diagram_ptr_->GetMutableSubsystemContext(
+        *lcm_parser_, &diagram_context);
+
+    // Wait for the first message.
+    drake::log()->info("Waiting for first lcm input message");
+    LcmHandleSubscriptionsUntil(drake_lcm_, [&]() {
+      return name_to_input_sub_map_.at(active_channel_).count() > 0;
+    });
+
+    // Initialize the context based on the first message.
+    const double t0 =
+        name_to_input_sub_map_.at(active_channel_).message().utime * 1e-6;
+    diagram_context.SetTime(t0);
+    auto& input_value = lcm_parser_->get_input_port(0).FixValue(
+        &lcm_parser_context,
+        name_to_input_sub_map_.at(active_channel_).message());
+
+    // "Simulator" time
+    double time = 0;  // initialize the current time with 0
+    double message_time;
+
+    drake::log()->info(diagram_name_ + " started");
+    // Run the simulation until end_time
+    while (time < end_time) {
+      // Update the name of the active channel if there are multiple inputs
+      if (is_multiple_inputs_) {
```

Comment:
Can this just check `switch_sub_ == nullptr`?

Reply:
Yes, it's done